flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaRecordSerializationSchemaBuilder.java

```diff
@@ -0,0 +1,270 @@
+     */
+    public KafkaRecordSerializationSchemaBuilder<IN> withPartitioner(
+            FlinkKafkaPartitioner<IN> partitioner) {
+        this.partitioner = checkNotNull(partitioner);
+        return this;
+    }
+
+    /**
+     * Sets a {@link SerializationSchema} which is used to serialize the incoming element to the key
+     * of the {@link ProducerRecord}.
+     *
+     * @param keySerializationSchema
+     * @return {@link KafkaRecordSerializationSchemaBuilder}
+     */
+    public KafkaRecordSerializationSchemaBuilder<IN> withKeySerializationSchema(
+            SerializationSchema<IN> keySerializationSchema) {
+        checkState(this.keySerializationSchema == null, exceedsKeySerializerConfigurations());
+        this.keySerializationSchema = checkNotNull(keySerializationSchema);
+        return this;
+    }
+
+    /**
+     * Sets Kafka's {@link Serializer} to serialize incoming elements to the key of the {@link
+     * ProducerRecord}.
+     *
+     * @param keySerializer
+     * @return {@link KafkaRecordSerializationSchemaBuilder}
+     */
+    public KafkaRecordSerializationSchemaBuilder<IN> withKeySerializer(
+            Class<? extends Serializer<IN>> keySerializer) {
+        checkState(keySerializationSchema == null, exceedsKeySerializerConfigurations());
+        this.keySerializationSchema =
+                new KafkaSerializerWrapper<>(keySerializer, topic, topicSelector);
+        return this;
+    }
+
+    /**
+     * Sets a configurable Kafka {@link Serializer} and pass a configuration to serialize incoming
+     * elements to the key of the {@link ProducerRecord}.
+     *
+     * @param keySerializerWithConfiguration
+     * @param configuration
+     * @param <T> type of the used serializer class
+     * @return {@link KafkaRecordSerializationSchemaBuilder}
+     */
+    public <T extends Configurable & Serializer<IN>>
+            KafkaRecordSerializationSchemaBuilder<IN> withConfigurableKeySerializer(
+                    Class<T> keySerializerWithConfiguration, Map<String, String> configuration) {
+        checkState(keySerializationSchema == null, exceedsKeySerializerConfigurations());
+        this.keySerializationSchema =
+                new KafkaSerializerWrapper<>(
+                        keySerializerWithConfiguration, configuration, topic, topicSelector);
+        return this;
+    }
+
+    /**
+     * Sets a {@link SerializationSchema} which is used to serialize the incoming element to the
+     * value of the {@link ProducerRecord}.
+     *
+     * @param valueSerializationSchema
+     * @return {@link KafkaRecordSerializationSchemaBuilder}
+     */
+    public KafkaRecordSerializationSchemaBuilder<IN> withValueSerializationSchema(
+            SerializationSchema<IN> valueSerializationSchema) {
+        checkState(this.valueSerializationSchema == null, exceedsValueSerializerConfigurations());
+        this.valueSerializationSchema = checkNotNull(valueSerializationSchema);
+        return this;
+    }
```

Comment:
There is a trick here to avoid binding the type of the KafkaRecordSerializationSchemaBuilder at `builder(...)`.
```suggestion
     public <T extends IN> KafkaRecordSerializationSchemaBuilder<T> withValueSerializationSchema(
            SerializationSchema<T> valueSerializationSchema)
        checkState(this.valueSerializationSchema == null, exceedsValueSerializerConfigurations());
        @SuppressWarning("unchecked")
        KafkaRecordSerializationSchemaBuilder<T> self = (KafkaRecordSerializationSchemaBuilder<T> ) this;
        self.valueSerializationSchema = checkNotNull(valueSerializationSchema);
        return self;
    }
```

Now we capture T and use that the main type. This also works of you still specify the type explicitly in `builder`.

Reply:
What kind of problem are you trying to solve with this suggestion?

Reply:
The ugliness that is called explicit type parameter in a method call. You can't tell me that you enjoy invoking a method `ClassA.<type>methodB` and prefer it over `ClassA.methodB`.

Reply:
I somewhat agree with your example but first I cannot reproduce my idea (it also work with only specifying `ClassA.methodB()` and second I do not think that it is worth the effort to complicate the code significantly for an "issue" probably most Java developers are used to see.

Reply:
I strongly disagree. Look at the user mailing list how often users have issues with `WatermarkStrategy` because only a small fraction of Java developers know that you can explicitly specify type args in method calls. Even fewer get the syntax correctly (why isn't it `class.method<type>(...)`)?
Our goal is to make ramp up as easy as possible. Every user that is not confused by that is well worth the 2 LOC "complexity".